ztest/ztest.go

```diff
@@ -271,12 +367,173 @@ func FromYAMLFile(filename string) (*ZTest, error) {
+		A:        difflib.SplitLines(expected),
+		FromFile: "expected",
+		B:        difflib.SplitLines(actual),
+		ToFile:   "actual",
+		Context:  5,
+	})
+	return fmt.Errorf("expected and actual outputs differ:\n%s", diff)
+}
+
+func run(t *testing.T, testname, bindir, dirname, filename, zq string, zt *ZTest) {
+	if err := zt.check(); err != nil {
+		t.Fatalf("%s: bad yaml format: %s", filename, err)
+	}
+	if zt.Script != "" {
+		if runtime.GOOS == "windows" {
+			t.Skip("skipping on windows")
+		}
+		adir, _ := filepath.Abs(dirname)
+		err := runsh(testname, bindir, adir, zt)
+		if err != nil {
+			t.Fatalf("%s: %s", filename, err)
+		}
+		return
+	}
+	out, errout, err := runzq(zq, zt.ZQL, zt.OutputFormat, zt.OutputFlags, zt.Input...)
+	if err != nil {
+		if zt.errRegex != nil {
+			if !zt.errRegex.Match([]byte(errout)) {
+				t.Fatalf("%s: error doesn't match expected error regex: %s %s", filename, zt.ErrorRE, errout)
+			}
+		} else {
+			if out != "" {
+				out = "\noutput:\n" + out
+			}
+			t.Fatalf("%s: %s%s", filename, err, out)
+		}
+	} else if zt.errRegex != nil {
+		t.Fatalf("%s: no error when expecting error regex: %s", filename, zt.ErrorRE)
+	}
+	expectedOut, oerr := zt.getOutput()
+	require.NoError(t, oerr)
+	if out != expectedOut {
+		a := expectedOut
+		b := out
+
+		if !utf8.ValidString(a) {
+			a = encodeHex(a)
+			b = encodeHex(b)
+		}
+
+		err := diffErr(a, b)
+		t.Fatalf("%s: expected and actual outputs differ:\n%s", filename, err)
+	}
+	if err == nil && errout != zt.Warnings {
+		err := diffErr(zt.Warnings, errout)
+		t.Fatalf("%s: expected and actual warnings differ:\n%s", filename, err)
+	}
+}
+
+func checkPatterns(patterns map[string]*Regexp, dir *Dir, stdout, stderr string) error {
+	for name, re := range patterns {
+		var body []byte
+		switch name {
+		case "stdout":
+			body = []byte(stdout)
+		case "stderr":
+			body = []byte(stderr)
+		default:
+			var err error
+			body, err = dir.Read(name)
+			if err != nil {
+				return fmt.Errorf("%s: %s", name, err)
+			}
+		}
+		if !re.Regexp.Match(body) {
+			return fmt.Errorf("regex mismatch: %s %s", re.Pattern, string(body))
+		}
+	}
+	return nil
+}
+
+func checkData(files map[string][]byte, dir *Dir, stdout, stderr string) error {
+	for name, expected := range files {
+		var actual []byte
+		switch name {
+		case "stdout":
+			actual = []byte(stdout)
+		case "stderr":
+			actual = []byte(stderr)
+		default:
+			var err error
+			actual, err = dir.Read(name)
+			if err != nil {
+				return fmt.Errorf("%s: %s", name, err)
+			}
+		}
+		if !bytes.Equal(expected, actual) {
+			return diffErr(string(expected), string(actual))
+		}
+	}
+	return nil
+}
+
+func runsh(testname, bindir, dirname string, zt *ZTest) error {
+	dir, err := NewDir(testname, dirname)
+	if err != nil {
+		return err
+	}
+	defer dir.RemoveAll()
+	for _, f := range zt.Inputs {
+		if f.Name == "stdin" {
+			return errors.New("cannot use stdin in a script test")
```

Comment:
The example in the package documentation above uses stdin; if it's not supported, the doc example should change.

Reply:
Oops, just pushed a fix and test for stdin.